packages/remix-server-runtime/__tests__/formData-test.ts

```diff
@@ -0,0 +1,175 @@
+  constructor() {
+    super("test error");
+  }
+}
+
+describe("parseMultipartFormData", () => {
+  it("can use a custom upload handler", async () => {
+    let formData = new NodeFormData();
+    formData.set("a", "value");
+    formData.set("blob", new Blob(["blob".repeat(1000)]), "blob.txt");
+    formData.set("file", new File(["file".repeat(1000)], "file.txt"));
+
+    let req = new NodeRequest("https://test.com", {
+      method: "post",
+      body: formData,
+    });
+
+    let parsedFormData = await parseMultipartFormData(
+      req,
+      async ({ filename, data, contentType }) => {
+        let chunks = [];
+        for await (let chunk of data) {
+          chunks.push(chunk);
+        }
+        if (filename) {
+          return new File(chunks, filename, { type: contentType });
+        }
+
+        return await new Blob(chunks, { type: contentType }).text();
+      }
+    );
+
+    expect(parsedFormData.get("a")).toBe("value");
+    let blob = parsedFormData.get("blob") as Blob;
+    expect(await blob.text()).toBe("blob".repeat(1000));
+    let file = parsedFormData.get("file") as File;
+    expect(file.name).toBe("file.txt");
+    expect(await file.text()).toBe("file".repeat(1000));
+  });
+
+  it("can return undefined", async () => {
+    let formData = new NodeFormData();
+    formData.set("a", "value");
+    formData.set("blob", new Blob(["blob".repeat(1000)]), "blob.txt");
+    formData.set("file", new File(["file".repeat(1000)], "file.txt"));
+
+    let req = new NodeRequest("https://test.com", {
+      method: "post",
+      body: formData,
+    });
+
+    let parsedFormData = await parseMultipartFormData(
+      req,
+      async () => undefined
+    );
+
+    expect(parsedFormData.get("a")).toBe(null);
+    expect(parsedFormData.get("blob")).toBe(null);
+    expect(parsedFormData.get("file")).toBe(null);
+  });
+
+  it("can throw errors in upload handlers", async () => {
+    let formData = new NodeFormData();
+    formData.set("blob", new Blob(["blob"]), "blob.txt");
+
+    let req = new NodeRequest("https://test.com", {
+      method: "post",
+      body: formData,
+    });
+
+    let error: Error;
+    try {
+      await parseMultipartFormData(req, async () => {
+        throw new CustomError();
+      });
+      throw new Error("should have thrown");
+    } catch (err) {
+      error = err;
+    }
+    expect(error).toBeInstanceOf(CustomError);
+    expect(error.message).toBe("test error");
+  });
+
+  describe("stream should propagate events", () => {
+    it("when controller errors", async () => {
+      let formData = new NodeFormData();
+      formData.set("a", "value");
+      formData.set("blob", new Blob(["blob".repeat(1000)]), "blob.txt");
+      formData.set("file", new File(["file".repeat(1000)], "file.txt"));
+
+      let underlyingRequest = new NodeRequest("https://test.com", {
+        method: "post",
+        body: formData,
+      });
+      let underlyingBody = await underlyingRequest.text();
```

Comment:
Are you running this through a `new Request` just so you can get a URL-encoded string out? Because if you are you can just use `new URLSearchParams(formData).toString()` to get the same thing.

Reply:
It's actually a multipart formdata body. It's to test what happens when things are interrupted.